pkg/kubelet/kubelet_node_status_test.go

```diff
@@ -1580,3 +1584,78 @@ func TestValidateNodeIPParam(t *testing.T) {
 		}
 	}
 }
+
+func TestRegisterWithApiServerWithTaint(t *testing.T) {
+	testKubelet := newTestKubelet(t, false /* controllerAttachDetachEnabled */)
+	defer testKubelet.Cleanup()
+	kubelet := testKubelet.kubelet
+	kubeClient := testKubelet.fakeKubeClient
+
+	machineInfo := &cadvisorapi.MachineInfo{
+		MachineID:      "123",
+		SystemUUID:     "abc",
+		BootID:         "1b3",
+		NumCores:       2,
+		MemoryCapacity: 1024,
+	}
+	mockCadvisor := testKubelet.fakeCadvisor
+	mockCadvisor.On("MachineInfo").Return(machineInfo, nil)
+	versionInfo := &cadvisorapi.VersionInfo{
+		KernelVersion:      "3.16.0-0.bpo.4-amd64",
+		ContainerOsVersion: "Debian GNU/Linux 7 (wheezy)",
+		DockerVersion:      "1.5.0",
+	}
+	mockCadvisor.On("VersionInfo").Return(versionInfo, nil)
+	mockCadvisor.On("ImagesFsInfo").Return(cadvisorapiv2.FsInfo{
+		Usage:     400,
+		Capacity:  1000,
+		Available: 600,
+	}, nil)
+	mockCadvisor.On("RootFsInfo").Return(cadvisorapiv2.FsInfo{
+		Usage:    9,
+		Capacity: 10,
+	}, nil)
+	kubelet.machineInfo = machineInfo
+
+	var gotNode runtime.Object
+	kubeClient.AddReactor("create", "nodes", func(action core.Action) (bool, runtime.Object, error) {
+		createAction := action.(core.CreateAction)
+		gotNode = createAction.GetObject()
+		return true, gotNode, nil
+	})
+	kubeClient.AddReactor("*", "*", func(action core.Action) (bool, runtime.Object, error) {
+		return true, nil, fmt.Errorf("no reaction implemented for %s", action)
+	})
+
+	// Make node to be unschedulable.
+	kubelet.registerSchedulable = false
+
+	forEachFeatureGate(t, []utilfeature.Feature{features.TaintNodesByCondition}, func(t *testing.T) {
+		done := make(chan struct{})
+
+		// Reset kubelet status for each test.
+		kubelet.registrationCompleted = false
+
+		go func() {
+			kubelet.registerWithAPIServer()
+			done <- struct{}{}
+		}()
+		select {
+		case <-time.After(wait.ForeverTestTimeout):
```

Comment:
Is it likely that `registerWithAPIServer` will hang? If not I'd say let the test framework handle the timeout and avoid the complexity.

Reply:
should not hang, removed :)